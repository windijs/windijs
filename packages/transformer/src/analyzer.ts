import type { CallExpressionArgument, LeftHandSideExpression } from "@typescript-eslint/types/dist/generated/ast-spec";

import { parse } from "@typescript-eslint/typescript-estree";

export type Handler = {
  type: string,
  guard?: string,
  arguments: CallExpressionArgument[],
}

let HANDLERS: Handler[] = [];

function analyzeHandler (expr: LeftHandSideExpression) {
  if (expr.type === "CallExpression" && expr.callee.type === "MemberExpression" && expr.callee.property.type === "Identifier") {
    analyzeHandler(expr.callee.object);
    switch (expr.callee.property.name) {
    case "use":
      for (const arg of expr.arguments) {
        if (arg.type === "CallExpression" && arg.callee.type === "Identifier" && arg.callee.name.endsWith("Handler")) {
          HANDLERS.push({ type: arg.callee.name, arguments: arg.arguments });
        }
      }
      break;
    case "case":
      if (expr.arguments[0].type === "Literal" && typeof expr.arguments[0].value === "string" && expr.arguments[1].type === "CallExpression" && expr.arguments[1].callee.type === "Identifier" && expr.arguments[1].callee.name.endsWith("Handler")) {
        HANDLERS.push({ type: expr.arguments[1].callee.name, arguments: expr.arguments[1].arguments, guard: expr.arguments[0].value });
      }
      break;
    }
  }
}

/**
 * Analyze handlers using in utilities, the results can be used to quickly generate .dts file.
 * @param code JavaScript/TypeScript code
 * @returns Utility name and handlers Object
 */
export function analyzeUtilities (code: string) {
  HANDLERS = [];

  const ast = parse(code, { loc: false, range: false });

  let name: string;
  const utilities: Record<string, Handler[]> = {};
  for (const node of ast.body) {
    if (node.type === "ExportNamedDeclaration" && node.declaration?.type === "VariableDeclaration") {
      for (const decl of node.declaration.declarations) {
        if (decl.type === "VariableDeclarator" && decl.id.type === "Identifier") {
          name = decl.id.name;
          HANDLERS = [];
          if (decl.init?.type === "CallExpression" && decl.init.callee.type === "MemberExpression" && decl.init.callee.property.type === "Identifier" && decl.init.callee.property.name === "init") {
            analyzeHandler(decl.init.callee.object);
          }
          utilities[name] = HANDLERS;
        }
      }
    }
  }
  return utilities;
}
