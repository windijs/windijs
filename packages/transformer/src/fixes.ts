import { Bundle, Node, SourceFile, SyntaxKind, TransformerFactory, factory, isIdentifier, isTypeAliasDeclaration, visitEachChild, visitNode } from "typescript";

/**
 * Typescript Transform Plugin.
 *
 * when rollup-plugin-ts generate .dts file, it will ignore default value parameter for type declaration.
 * this transformer fixed this.
 */
export const fixHelperStyle: TransformerFactory<Bundle | SourceFile> = context => {
  return sourceFile => {
    const visitor = (node: Node): Node => {
      if (isTypeAliasDeclaration(node) && isIdentifier(node.name) && node.name.escapedText === "StyleObject") {
        return factory.updateTypeAliasDeclaration(node, node.decorators, node.modifiers, node.name, [factory.createTypeParameterDeclaration(undefined, "T", undefined, factory.createTypeLiteralNode([]))], node.type);
      }

      if (isTypeAliasDeclaration(node) && isIdentifier(node.name) && node.name.escapedText === "StyleProxy") {
        return factory.updateTypeAliasDeclaration(node, node.decorators, node.modifiers, node.name, [factory.createTypeParameterDeclaration(undefined, "T"), factory.createTypeParameterDeclaration(undefined, "O", undefined, factory.createTypeLiteralNode([]))], node.type);
      }

      if (isTypeAliasDeclaration(node) && isIdentifier(node.name) && node.name.escapedText === "BuildFunc") {
        return factory.updateTypeAliasDeclaration(node, node.decorators, node.modifiers, node.name, [factory.createTypeParameterDeclaration(undefined, "V", undefined, factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword)), factory.createTypeParameterDeclaration(undefined, "O", factory.createKeywordTypeNode(SyntaxKind.ObjectKeyword), factory.createTypeLiteralNode([]))], node.type);
      }

      return visitEachChild(node, visitor, context);
    };
    return visitNode(sourceFile, visitor);
  };
};

// export const fixTuplifyUnion: TransformerFactory<Bundle | SourceFile> = context => {
//   return sourceFile => {
//     const visitor = (node: Node): Node => {
//       if (isTypeAliasDeclaration(node) && isIdentifier(node.name) && node.name.escapedText === "TuplifyUnion") {
//         return factory.updateTypeAliasDeclaration(node, node.decorators, node.modifiers, node.name, [
//           factory.createTypeParameterDeclaration(undefined, "T"),
//           factory.createTypeParameterDeclaration(undefined, "L", undefined, factory.createTypeReferenceNode("LastOf", [factory.createTypeReferenceNode("T")])),
//           factory.createTypeParameterDeclaration(undefined, "N", undefined, factory.createConditionalTypeNode(factory.createTupleTypeNode([factory.createTypeReferenceNode("T")]), factory.createTupleTypeNode([factory.createKeywordTypeNode(SyntaxKind.NeverKeyword)]), factory.createLiteralTypeNode(factory.createTrue()), factory.createLiteralTypeNode(factory.createFalse()))),
//         ], node.type);
//       }
//       return visitEachChild(node, visitor, context);
//     };
//     return visitNode(sourceFile, visitor);
//   };
// };
