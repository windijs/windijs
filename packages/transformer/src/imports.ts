import { Bundle, Node, SourceFile, TransformerFactory, factory, isIdentifier, isImportDeclaration, isImportTypeNode, isLiteralTypeNode, isNamedImports, isStringLiteral, visitEachChild, visitNode } from "typescript";

// TODO: we switched our dts plugin, maybe this transformer no longer needed.

/**
 * Typescript Transform Plugin.
 *
 * when rollup-plugin-ts generate .dts file, it will leave some extras, for example `import("@windi/helper").StyleObject`, even when `StyleObject` is already imported.
 * this transformer fixed this.
 */
export const importsTransformer: TransformerFactory<Bundle | SourceFile> = context => {
  return sourceFile => {
    const imports: Record<string, string[]> = {};
    const visitor = (node: Node): Node => {
      if (isImportDeclaration(node) && isStringLiteral(node.moduleSpecifier) && node.importClause?.namedBindings) {
        if (isNamedImports(node.importClause.namedBindings)) {
          imports[node.moduleSpecifier.text] = node.importClause.namedBindings.elements.map(i => i.name.escapedText.toString());
        }
      }

      if (isImportTypeNode(node)) {
        if (isLiteralTypeNode(node.argument) && isStringLiteral(node.argument.literal) && node.argument.literal.text in imports && node.qualifier && isIdentifier(node.qualifier) && imports[node.argument.literal.text].includes(node.qualifier.escapedText.toString())) {
          return factory.createTypeReferenceNode(node.qualifier, node.typeArguments);
        }
      }
      return visitEachChild(node, visitor, context);
    };
    return visitNode(sourceFile, visitor);
  };
};

/**
 * Typescript Transform Plugin.
 *
 * Change all Dynamic Imports to Global Imports. For Example, `import("@windi/helpers").StyleObject<{}>` -> `StyleObject<{}>`.
 */
export const importTypesTransformer: TransformerFactory<Bundle | SourceFile> = context => {
  return sourceFile => {
    const imports: Record<string, string[]> = {};
    const visitor = (node: Node): Node => {
      let item: string;
      if (isImportTypeNode(node)) {
        if (isLiteralTypeNode(node.argument) && isStringLiteral(node.argument.literal) && node.qualifier && isIdentifier(node.qualifier)) {
          item = node.qualifier.escapedText.toString();
          if (!(node.argument.literal.text in imports)) {
            imports[node.argument.literal.text] = [item];
          } else if (!imports[node.argument.literal.text].includes(item)) {
            imports[node.argument.literal.text].push(item);
          }
          return factory.createTypeReferenceNode(node.qualifier, node.typeArguments);
        }
      }
      return visitEachChild(node, visitor, context);
    };
    const transformed = visitNode(sourceFile, visitor) as SourceFile;
    return factory.updateSourceFile(transformed, [
      ...Object.entries(imports).map(([k, v]) => factory.createImportDeclaration(undefined, undefined, factory.createImportClause(false, undefined, factory.createNamedImports(v.map(i => factory.createImportSpecifier(false, undefined, factory.createIdentifier(i))))), factory.createStringLiteral(k))),
      ...transformed.statements,
    ]);
  };
};
