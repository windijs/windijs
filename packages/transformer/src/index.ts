export * from "./imports";
export * from "./utility";
export * from "./analyzer";
export * from "./intersection";
export * from "./fixes";
export * from "./utils";
