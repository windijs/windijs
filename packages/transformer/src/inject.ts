import { Bundle, SourceFile, SyntaxKind, TransformerFactory, Visitor, factory, isIdentifier, isImportDeclaration, isLiteralTypeNode, isStringLiteral, isTypeReferenceNode, visitEachChild, visitNode } from "typescript";

/**
 * Typescript Transform Plugin.
 *
 * Convert `Pick<..., "$...">` to `Inject<..., "$...">`, also add the Inject type.
 */
export const injectTransformer: TransformerFactory<Bundle | SourceFile> = context => {
  return sourceFile => {
    const visitor: Visitor = (node) => {
      if (isTypeReferenceNode(node) && isIdentifier(node.typeName) && node.typeName.escapedText === "Pick" && node.typeArguments && isLiteralTypeNode(node.typeArguments[1]) && isStringLiteral(node.typeArguments[1].literal) && node.typeArguments[1].literal.text.startsWith("$windi.")) {
        return factory.updateTypeReferenceNode(node, factory.createIdentifier("Inject"), node.typeArguments);
      }
      return visitEachChild(node, visitor, context);
    };
    const transformed = visitNode(sourceFile, visitor) as SourceFile;
    return factory.updateSourceFile(transformed, [
      ...transformed.statements.filter(i => isImportDeclaration(i)),
      factory.createTypeAliasDeclaration(undefined, undefined, "Inject", [factory.createTypeParameterDeclaration(undefined, "A"), factory.createTypeParameterDeclaration(undefined, "B")], factory.createConditionalTypeNode(factory.createTypeReferenceNode("B"), factory.createKeywordTypeNode(SyntaxKind.ObjectKeyword), factory.createTypeReferenceNode("B"), factory.createTypeReferenceNode("A"))),
      ...transformed.statements.filter(i => !isImportDeclaration(i)),
    ]);
  };
};
