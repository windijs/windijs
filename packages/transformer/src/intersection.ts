import { Bundle, Node, SourceFile, TransformerFactory, factory, isIndexSignatureDeclaration, isIntersectionTypeNode, isTypeLiteralNode, visitEachChild, visitNode } from "typescript";

/**
 * Typescript Transform Plugin.
 *
 * This transformer combine seperated object type in intersection type. For Example
 *
 * ```ts
 * { a: string } & { b: string } -> { a: string; b: string; }
 * ```
 */
export const intersectionTransformer: TransformerFactory<Bundle | SourceFile> = context => {
  return sourceFile => {
    const visitor = (node: Node): Node => {
      if (isIntersectionTypeNode(node)) {
        const literals = [];
        const newTypes = [];
        for (const type of node.types) {
          if (isTypeLiteralNode(type) && !type.members.find(i => isIndexSignatureDeclaration(i))) {
            literals.push(...type.members);
          } else {
            newTypes.push(type);
          }
        }

        if (literals.length > 0) newTypes.unshift(factory.createTypeLiteralNode(literals));

        return factory.createIntersectionTypeNode(newTypes);
      }

      return visitEachChild(node, visitor, context);
    };

    return visitNode(sourceFile, visitor);
  };
};
