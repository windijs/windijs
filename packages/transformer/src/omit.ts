import {
  Bundle,
  Node,
  SourceFile,
  TransformerFactory,
  TypeElement,
  factory,
  isIdentifier,
  isLiteralTypeNode,
  isPropertySignature,
  isStringLiteral,
  isTypeLiteralNode,
  isTypeReferenceNode,
  isUnionTypeNode,
  visitEachChild,
  visitNode,
} from "typescript";

/**
 * Typescript Transform Plugin.
 *
 * Remove useless Omit, for example, `Omit<{none: StyleObject<{}>}, "DEFAULT">` -> `{none: StyleObject<{}>}`
 */
export const omitTransformer: TransformerFactory<Bundle | SourceFile> = context => {
  return sourceFile => {
    const visitor = (node: Node): Node => {
      if (isTypeReferenceNode(node) && isIdentifier(node.typeName) && node.typeName.escapedText === "Omit")
        if (node.typeArguments) {
          const value = node.typeArguments[0];
          const omitted = node.typeArguments[1];
          const keys: string[] = isStringLiteral(omitted)
            ? [omitted.text]
            : isUnionTypeNode(omitted)
            ? omitted.types.map(i => (isLiteralTypeNode(i) && isStringLiteral(i.literal) ? i.literal.text : ""))
            : [];
          if (isTypeLiteralNode(value)) {
            const newMembers: TypeElement[] = [];
            for (const m of value.members) {
              if (isPropertySignature(m)) {
                if (isStringLiteral(m.name) && keys.includes(m.name.text)) continue;
                if (isIdentifier(m.name) && keys.includes(m.name.escapedText.toString())) continue;
              }
              newMembers.push(m);
            }
            return factory.createTypeLiteralNode(newMembers);
          }
        }

      return visitEachChild(node, visitor, context);
    };
    return visitNode(sourceFile, visitor);
  };
};
