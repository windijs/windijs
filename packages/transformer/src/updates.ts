import {
  Bundle,
  Node,
  SourceFile,
  TransformerFactory,
  TypeNode,
  factory,
  isIdentifier,
  isVariableDeclaration,
  visitEachChild,
  visitNode,
} from "typescript";

/**
 * Typescript Transform Plugin.
 * Transform variable type to another type
 */
export function updateVariableType(dict: Record<string, (node: TypeNode) => TypeNode>) {
  const constsTransformer: TransformerFactory<Bundle | SourceFile> = context => {
    return sourceFile => {
      const visitor = (node: Node): Node => {
        if (isVariableDeclaration(node) && isIdentifier(node.name) && node.type) {
          const key = node.name.escapedText.toString();
          if (key in dict) return factory.updateVariableDeclaration(node, node.name, node.exclamationToken, dict[key](node.type), node.initializer);
        }
        return visitEachChild(node, visitor, context);
      };
      return visitNode(sourceFile, visitor);
    };
  };
  return constsTransformer;
}
