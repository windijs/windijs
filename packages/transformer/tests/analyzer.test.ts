import { analyzeUtilities } from "../src";

test("Analyze single utility", () => {
  const code = `
export const bg = createUtility("bg")
  .use(colorHandler(colors, "backgroundColor", "--w-bg-opacity"))
  .use(configHandler(backgroundAttachmentConfig, "backgroundAttachment"))
  .use(configHandler(backgroundPositionConfig, "backgroundPosition"))
  .use(configHandler(backgroundRepeatConfig, "backgroundRepeat"))
  .use(configHandler(backgroundSizeConfig, "backgroundSize"))
  .use(configHandler(backgroundImageConfig, "backgroundImage"))
  .case("clip", configHandler(backgroundClipConfig, ["backgroundClip", prop\`-webkit-background-clip\`]))
  .case("blend", configHandler(blendModeConfig, "backgroundBlendMode"))
  .case("origin", configHandler(backgroundOriginConfig, "backgroundOrigin"))
  .case("opacity", configHandler(opacityConfig, prop\`--w-bg-opacity\`))
  .case("gradient", callHandler(buildLinearGradient, meld(
    configHandler(gradientDirectionConfig, buildGradientDirection),
    configHandler(gradientConfig, "backgroundImage"),
  )))
  .init();
  `;

  expect(analyzeUtilities(code)).toMatchSnapshot();
});

test("Analyze multi utilities", () => {
  const code = `
export const shadow = createUtility("shadow")
  .use(configHandler(boxShadowConfig, buildBoxShadowSize))
  .use(colorHandler(colors, buildBoxShadowColor))
  .case("opacity", configHandler(opacityConfig, prop\`--w-shadow-color-opacity\`))
  .init();

export const opacity = createUtility("opacity")
  .use(configHandler(opacityConfig, "opacity"))
  .init();

export const blend = createUtility("blend")
  .use(configHandler(blendModeConfig, "mixBlendMode"))
  .init();
  `;

  expect(analyzeUtilities(code)).toMatchSnapshot();
});
