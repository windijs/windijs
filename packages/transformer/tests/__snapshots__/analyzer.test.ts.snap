// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyze multi utilities 1`] = `
Object {
  "blend": Array [
    Object {
      "arguments": Array [
        Object {
          "name": "blendModeConfig",
          "type": "Identifier",
        },
        Object {
          "raw": "\\"mixBlendMode\\"",
          "type": "Literal",
          "value": "mixBlendMode",
        },
      ],
      "type": "configHandler",
    },
  ],
  "opacity": Array [
    Object {
      "arguments": Array [
        Object {
          "name": "opacityConfig",
          "type": "Identifier",
        },
        Object {
          "raw": "\\"opacity\\"",
          "type": "Literal",
          "value": "opacity",
        },
      ],
      "type": "configHandler",
    },
  ],
  "shadow": Array [
    Object {
      "arguments": Array [
        Object {
          "name": "boxShadowConfig",
          "type": "Identifier",
        },
        Object {
          "name": "buildBoxShadowSize",
          "type": "Identifier",
        },
      ],
      "type": "configHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "colors",
          "type": "Identifier",
        },
        Object {
          "name": "buildBoxShadowColor",
          "type": "Identifier",
        },
      ],
      "type": "colorHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "opacityConfig",
          "type": "Identifier",
        },
        Object {
          "quasi": Object {
            "expressions": Array [],
            "quasis": Array [
              Object {
                "tail": true,
                "type": "TemplateElement",
                "value": Object {
                  "cooked": "--w-shadow-color-opacity",
                  "raw": "--w-shadow-color-opacity",
                },
              },
            ],
            "type": "TemplateLiteral",
          },
          "tag": Object {
            "name": "prop",
            "type": "Identifier",
          },
          "type": "TaggedTemplateExpression",
          "typeParameters": undefined,
        },
      ],
      "guard": "opacity",
      "type": "configHandler",
    },
  ],
}
`;

exports[`Analyze single utility 1`] = `
Object {
  "bg": Array [
    Object {
      "arguments": Array [
        Object {
          "name": "colors",
          "type": "Identifier",
        },
        Object {
          "raw": "\\"backgroundColor\\"",
          "type": "Literal",
          "value": "backgroundColor",
        },
        Object {
          "raw": "\\"--w-bg-opacity\\"",
          "type": "Literal",
          "value": "--w-bg-opacity",
        },
      ],
      "type": "colorHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "backgroundAttachmentConfig",
          "type": "Identifier",
        },
        Object {
          "raw": "\\"backgroundAttachment\\"",
          "type": "Literal",
          "value": "backgroundAttachment",
        },
      ],
      "type": "configHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "backgroundPositionConfig",
          "type": "Identifier",
        },
        Object {
          "raw": "\\"backgroundPosition\\"",
          "type": "Literal",
          "value": "backgroundPosition",
        },
      ],
      "type": "configHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "backgroundRepeatConfig",
          "type": "Identifier",
        },
        Object {
          "raw": "\\"backgroundRepeat\\"",
          "type": "Literal",
          "value": "backgroundRepeat",
        },
      ],
      "type": "configHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "backgroundSizeConfig",
          "type": "Identifier",
        },
        Object {
          "raw": "\\"backgroundSize\\"",
          "type": "Literal",
          "value": "backgroundSize",
        },
      ],
      "type": "configHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "backgroundImageConfig",
          "type": "Identifier",
        },
        Object {
          "raw": "\\"backgroundImage\\"",
          "type": "Literal",
          "value": "backgroundImage",
        },
      ],
      "type": "configHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "backgroundClipConfig",
          "type": "Identifier",
        },
        Object {
          "elements": Array [
            Object {
              "raw": "\\"backgroundClip\\"",
              "type": "Literal",
              "value": "backgroundClip",
            },
            Object {
              "quasi": Object {
                "expressions": Array [],
                "quasis": Array [
                  Object {
                    "tail": true,
                    "type": "TemplateElement",
                    "value": Object {
                      "cooked": "-webkit-background-clip",
                      "raw": "-webkit-background-clip",
                    },
                  },
                ],
                "type": "TemplateLiteral",
              },
              "tag": Object {
                "name": "prop",
                "type": "Identifier",
              },
              "type": "TaggedTemplateExpression",
              "typeParameters": undefined,
            },
          ],
          "type": "ArrayExpression",
        },
      ],
      "guard": "clip",
      "type": "configHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "blendModeConfig",
          "type": "Identifier",
        },
        Object {
          "raw": "\\"backgroundBlendMode\\"",
          "type": "Literal",
          "value": "backgroundBlendMode",
        },
      ],
      "guard": "blend",
      "type": "configHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "backgroundOriginConfig",
          "type": "Identifier",
        },
        Object {
          "raw": "\\"backgroundOrigin\\"",
          "type": "Literal",
          "value": "backgroundOrigin",
        },
      ],
      "guard": "origin",
      "type": "configHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "opacityConfig",
          "type": "Identifier",
        },
        Object {
          "quasi": Object {
            "expressions": Array [],
            "quasis": Array [
              Object {
                "tail": true,
                "type": "TemplateElement",
                "value": Object {
                  "cooked": "--w-bg-opacity",
                  "raw": "--w-bg-opacity",
                },
              },
            ],
            "type": "TemplateLiteral",
          },
          "tag": Object {
            "name": "prop",
            "type": "Identifier",
          },
          "type": "TaggedTemplateExpression",
          "typeParameters": undefined,
        },
      ],
      "guard": "opacity",
      "type": "configHandler",
    },
    Object {
      "arguments": Array [
        Object {
          "name": "buildLinearGradient",
          "type": "Identifier",
        },
        Object {
          "arguments": Array [
            Object {
              "arguments": Array [
                Object {
                  "name": "gradientDirectionConfig",
                  "type": "Identifier",
                },
                Object {
                  "name": "buildGradientDirection",
                  "type": "Identifier",
                },
              ],
              "callee": Object {
                "name": "configHandler",
                "type": "Identifier",
              },
              "optional": false,
              "type": "CallExpression",
            },
            Object {
              "arguments": Array [
                Object {
                  "name": "gradientConfig",
                  "type": "Identifier",
                },
                Object {
                  "raw": "\\"backgroundImage\\"",
                  "type": "Literal",
                  "value": "backgroundImage",
                },
              ],
              "callee": Object {
                "name": "configHandler",
                "type": "Identifier",
              },
              "optional": false,
              "type": "CallExpression",
            },
          ],
          "callee": Object {
            "name": "meld",
            "type": "Identifier",
          },
          "optional": false,
          "type": "CallExpression",
        },
      ],
      "guard": "gradient",
      "type": "callHandler",
    },
  ],
}
`;
